spring-cloud-aws-sns/src/main/java/io/awspring/cloud/sns/core/NotificationMessagingTemplate.java

```diff
@@ -0,0 +1,159 @@
+
+/**
+ * @author Alain Sahli
+ * @author Matej Nedic
+ * @since 1.0
+ */
+public class NotificationMessagingTemplate extends AbstractMessageSendingTemplate<TopicMessageChannel>
+		implements DestinationResolvingMessageSendingOperations<TopicMessageChannel> {
+
+	private final SnsClient snsClient;
+
+	private final DestinationResolver<String> destinationResolver;
+
+	public NotificationMessagingTemplate(SnsClient snsClient) {
+		this(snsClient, null);
+	}
+
+	public NotificationMessagingTemplate(SnsClient snsClient, MessageConverter messageConverter) {
+		this.destinationResolver = new DynamicTopicDestinationResolver(snsClient);
+		this.snsClient = snsClient;
+		initMessageConverter(messageConverter);
+	}
+
+	public NotificationMessagingTemplate(SnsClient snsClient, DestinationResolver<String> destinationResolver,
+			MessageConverter messageConverter) {
+		this.destinationResolver = destinationResolver;
+		this.snsClient = snsClient;
+		initMessageConverter(messageConverter);
+	}
+
+	public void setDefaultDestinationName(String defaultDestination) {
+		super.setDefaultDestination(resolveMessageChannelByLogicalName(defaultDestination));
+	}
+
+	@Override
+	protected void doSend(TopicMessageChannel destination, Message<?> message) {
+		destination.send(message);
+	}
+
+	@Override
+	public void send(String destinationName, Message<?> message) throws MessagingException {
+		TopicMessageChannel channel = resolveMessageChannelByLogicalName(destinationName);
+		doSend(channel, message);
+	}
+
+	@Override
+	public <T> void convertAndSend(String destinationName, T payload) throws MessagingException {
+		TopicMessageChannel channel = resolveMessageChannelByLogicalName(destinationName);
+		convertAndSend(channel, payload);
+	}
+
+	@Override
+	public <T> void convertAndSend(String destinationName, T payload, Map<String, Object> headers)
+			throws MessagingException {
+		TopicMessageChannel channel = resolveMessageChannelByLogicalName(destinationName);
+		convertAndSend(channel, payload, headers);
+	}
+
+	@Override
+	public <T> void convertAndSend(String destinationName, T payload, MessagePostProcessor postProcessor)
+			throws MessagingException {
+		TopicMessageChannel channel = resolveMessageChannelByLogicalName(destinationName);
+		convertAndSend(channel, payload, postProcessor);
+	}
+
+	@Override
+	public <T> void convertAndSend(String destinationName, T payload, Map<String, Object> headers,
+			MessagePostProcessor postProcessor) throws MessagingException {
+		TopicMessageChannel channel = resolveMessageChannelByLogicalName(destinationName);
+		convertAndSend(channel, payload, headers, postProcessor);
+	}
+
+	protected TopicMessageChannel resolveMessageChannelByLogicalName(String destination) {
+		String physicalResourceId = this.destinationResolver.resolveDestination(destination);
+		return resolveMessageChannel(physicalResourceId);
+	}
+
+	protected void initMessageConverter(MessageConverter messageConverter) {
```

Comment:
I think it would be more readable if this method is side effect free - instead of setting message converter should return composite message converter.